### PR TITLE
SAA-356: Add minimumIncentiveNomisCode

### PR DESCRIPTION
**Pull request by gerryallan-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt [hunk 1]
```diff
@@ -63,6 +63,8 @@ data class Activity(
 
   var riskLevel: String? = null,
 
+  var minimumIncentiveNomisCode: String? = null,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt:66]: This comment isn't related to your change just something I have noticed.

This wasn't part of the ticket @timharrison-moj. I think the incentive level fields should be mandatory here.

The same goes for activity pay, should be mandatory there as well.

Could we address this as well as part of this change?

**Review by matt-crowson-moj — DISMISSED** [review]
LGTM (**DO NOT MERGE**)

I have added a comment which is an addition so not sure yet on whether to do it.

Note this has a breaking change with the DB and the front end.  So it will require a DB drop and the front end changes merged altogether.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityCreateRequest.kt [hunk 2]
```diff
@@ -59,6 +59,10 @@ data class ActivityCreateRequest(
   @Schema(description = "The most recent risk assessment level for this activity", example = "High")
   val riskLevel: String?,
 
+  @field:Size(max = 10, message = "Minimum incentive level NOMIS code should not exceed {max} characters")
+  @Schema(description = "The NOMIS code for the minimum incentive/earned privilege level for this activity", example = "BAS")
+  val minimumIncentiveNomisCode: String?,
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityCreateRequest.kt:64]: Field size - think they're all 3 chars for nomis codes - but could check this in incentives-api ,which is now the master for them?

File: src/main/resources/migration/data/R__2_1__activities.sql [hunk 3]
```diff
@@ -1,7 +1,7 @@
...
-       (2, 'MDI', 1, 1, true, false, false, false, 'H', 'English level 1', 'A basic english course suitable for introduction to the subject', '2022-10-10', null, null, null, '2022-10-10 09:00:00', 'SEED USER'),
-       (3, 'MDI', 3, 1, true, false, false, false, 'H', 'Wing cleaning', 'Cleaning and upkeep of the wing', '2022-10-10', null, null, null, '2022-10-10 09:00:00', 'SEED USER'),
-       (4, 'MDI', 4, 1, false, false, false, false, 'H', 'Gym', 'Time for exercise', '2022-10-10', null, null, null, '2022-10-10 09:00:00', 'SEED USER');
+insert into activity(activity_id, prison_code, activity_category_id, activity_tier_id, attendance_required, in_cell, piece_work, outside_work, pay_per_session, summary, description, start_date, end_date, risk_level, minimum_incentive_nomis_code, minimum_incentive_level, created_time, created_by)
+values (1, 'MDI', 1, 1, true, false, false, false, 'H', 'Maths level 1', 'A basic maths course suitable for introduction to the subject', '2022-10-10', null, null, null, null, '2022-10-10 09:00:00', 'SEED USER'),
```
Comment (timharrison-moj) [inline src/main/resources/migration/data/R__2_1__activities.sql:2]: Probably always need a value one of BASE, STD, ENH, EN2.

File: src/test/resources/test_data/seed-activity-id-10.sql [hunk 4]
```diff
@@ -1,8 +1,8 @@
 --
 -- One activity with one schedule, one slot and one instance but no allocations
 --
-insert into activity(activity_id, prison_code, activity_category_id, activity_tier_id, attendance_required, in_cell, piece_work, outside_work, pay_per_session, summary, description, start_date, end_date, risk_level, minimum_incentive_level, created_time, created_by)
-values (10, 'MDI', 2, 2, true, false, false, false, 'H', 'Geography', 'Geography Level 1', '2022-10-01', null, null, null, '2022-9-21 00:00:00', 'SEED USER');
+insert into activity(activity_id, prison_code, activity_category_id, activity_tier_id, attendance_required, in_cell, piece_work, outside_work, pay_per_session, summary, description, start_date, end_date, risk_level, minimum_incentive_nomis_code, minimum_incentive_level, created_time, created_by)
+values (10, 'MDI', 2, 2, true, false, false, false, 'H', 'Geography', 'Geography Level 1', '2022-10-01', null, null, null, null, '2022-9-21 00:00:00', 'SEED USER');
```
Comment (timharrison-moj) [inline src/test/resources/test_data/seed-activity-id-10.sql:5]: Default to BAS, probably?

File: src/main/resources/migration/data/R__2_1__activities.sql (hunk 3 above)
Comment (gerryallan-moj) [reply src/main/resources/migration/data/R__2_1__activities.sql]: Default values added

File: src/test/resources/test_data/seed-activity-id-10.sql (hunk 4 above)
Comment (gerryallan-moj) [reply src/test/resources/test_data/seed-activity-id-10.sql]: Default value added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt [hunk 5]
```diff
@@ -8,8 +8,11 @@ data class ActivityPay(
   @Schema(description = "The internally-generated ID for this activity pay", example = "123456")
   val id: Long,
 
+  @Schema(description = "The NOMIS code for the incentive/earned privilege level (nullable)", example = "BAS")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt:11]: Minor point but need to remove the word nullable from incentive descriptions.  They are no longer nullable.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt [hunk 6]
```diff
@@ -63,7 +63,9 @@ data class Activity(
 
   var riskLevel: String? = null,
 
-  var minimumIncentiveLevel: String? = null,
+  var minimumIncentiveNomisCode: String,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt:66]: @timharrison-moj this is as per ticket but I wonder if we should also include 'level' in the name of the NOMIS fields as well? Same goes for activity pay.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityPayCreateRequest.kt [hunk 7]
```diff
@@ -1,16 +1,23 @@
...
 import jakarta.validation.constraints.NotNull
 import jakarta.validation.constraints.Positive
 import jakarta.validation.constraints.Size
 
 @Schema(description = "Describes the pay rates and bands to be created for an activity")
 data class ActivityPayCreateRequest(
 
+  @field:NotEmpty(message = "NOMIS code for the minimum incentive level must be supplied")
+  @field:Size(max = 3, message = "NOMIS code for the incentive level should not exceed {max} characters")
+  @Schema(description = "The NOMIS code for the incentive/earned privilege level (nullable)", example = "BAS")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityPayCreateRequest.kt:14]: Same here. Minor point but need to remove the word nullable from incentive descriptions. They are no longer nullable.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt (hunk 5 above)
Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/ActivityPay.kt]: updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityPayCreateRequest.kt (hunk 7 above)
Comment (gerryallan-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/request/ActivityPayCreateRequest.kt]: updated

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**